lib/ui/painting/fragment_shader.cc

```diff
@@ -0,0 +1,84 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <iostream>
+
+#include "flutter/lib/ui/painting/fragment_shader.h"
+
+#include "flutter/lib/ui/dart_wrapper.h"
+#include "flutter/lib/ui/ui_dart_state.h"
+#include "third_party/skia/include/core/SkString.h"
+#include "third_party/tonic/converter/dart_converter.h"
+#include "third_party/tonic/dart_args.h"
+#include "third_party/tonic/dart_binding_macros.h"
+#include "third_party/tonic/dart_library_natives.h"
+#include "third_party/tonic/typed_data/typed_list.h"
+
+using tonic::ToDart;
+
+namespace flutter {
+
+static void FragmentShader_constructor(Dart_NativeArguments args) {
+  DartCallConstructor(&FragmentShader::Create, args);
+}
+
+IMPLEMENT_WRAPPERTYPEINFO(ui, FragmentShader);
+
+#define FOR_EACH_BINDING(V) \
+  V(FragmentShader, init)   \
+  V(FragmentShader, update)
+
+FOR_EACH_BINDING(DART_NATIVE_CALLBACK)
+
+void FragmentShader::RegisterNatives(tonic::DartLibraryNatives* natives) {
+  natives->Register(
+      {{"FragmentShader_constructor", FragmentShader_constructor, 1, true},
+       FOR_EACH_BINDING(DART_REGISTER_NATIVE)});
+}
+
+sk_sp<SkShader> FragmentShader::shader(SkSamplingOptions sampling) {
+  // TODO(insert github issue): implement this method.
+  return shader_;
+}
+
+void FragmentShader::init(std::string sksl, bool debugPrintSksl) {
+  SkRuntimeEffect::Result result =
+      SkRuntimeEffect::MakeForShader(SkString(sksl));
+  runtime_effect_ = result.effect;
+
+  // TODO(reviewers): what is the proper way to throw and log these?
+  if (runtime_effect_ == nullptr) {
+    Dart_Handle err = Dart_ThrowException(tonic::ToDart(
+        std::string("Invalid SkSL:\n") + sksl.c_str() +
+        std::string("\nSkSL Error:\n") + result.errorText.c_str()));
+    if (err) {
+      FML_DLOG(ERROR) << Dart_GetError(err);
+    } else {
+      FML_DLOG(ERROR) << std::string("SkSL:\n") + sksl.c_str();
+    }
+    return;
```

Comment:
This block is only reachable if `Dart_ThrowException()` fails to throw an exception, e.g. when there are no Dart frames on the stack. That is probably a severe enough error that we'd want to abort the process.

Reply:
Good point, changed this to follow how `Dart_ThrowException()` is used elsewhere, such as canvas.cc